src/data/mv_polynomial/derivation.lean

```diff
@@ -0,0 +1,122 @@
+/-
+Copyright (c) 2021 Yury Kudryashov. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yury Kudryashov
+-/
+import data.mv_polynomial.basic
+import ring_theory.derivation
+
+/-!
+# Derivations of multivariate polynomials
+
+In this file we prove that a variation of `mv_polynomial σ R` is defined by its values on all
+monomials `mv_polynomial.X i`. We also provide a constructor `mv_polynomial.mk_derivation` that
+builds a derivation from its values on `X i`s and a linear equivalence
+`mv_polynomial.equiv_derivation` between `σ → A` and `derivation (mv_polynomial σ R) A`. -/
+
+namespace mv_polynomial
+
+open_locale big_operators
+
+noncomputable theory
+
+variables {σ R A : Type*} [comm_ring R] [add_comm_group A]
+  [module R A] [module (mv_polynomial σ R) A]
+
+section
+
+variable (R)
+
+def mk_derivationₗ (f : σ → A) :
+  mv_polynomial σ R →ₗ[R] A :=
+finsupp.lsum R $ λ xs : σ →₀ ℕ, (linear_map.ring_lmap_equiv_self R _ R).symm $
+  xs.sum $ λ i k, monomial (xs - finsupp.single i 1) (k : R) • f i
+
+end
+
+variables [is_scalar_tower R (mv_polynomial σ R) A]
+
+lemma leibniz_iff_X
+  (f : mv_polynomial σ R →ₗ[R] A) :
+  (∀ p q, f (p * q) = p • f q + q • f p) ↔ f 1 = 0 ∧
+    (∀ s i, f (monomial s 1 * X i) = (monomial s 1 : mv_polynomial σ R) • f (X i) +
+      (X i : mv_polynomial σ R) • f (monomial s 1)) :=
+begin
+  refine ⟨λ H, ⟨_, λ p i, H _ _⟩, λ H, _⟩,
+  { simpa using H 1 1 },
+  { have : ∀ p i, f (p * X i) = p • f (X i) + (X i : mv_polynomial σ R) • f p,
+    { intros p i,
+      induction p using mv_polynomial.induction_on' with s r p q hp hq,
+      { rw [← mul_one r, ← C_mul_monomial, mul_assoc, C_mul', f.map_smul, H.2, C_mul', smul_assoc,
+          smul_add, f.map_smul, smul_comm r (X i)], apply_instance },
+      { rw [add_mul, f.map_add, f.map_add, hp, hq, add_smul, smul_add, add_add_add_comm] } },
+    intros p q,
+    induction q using mv_polynomial.induction_on,
+    case h_C : c {
+      rw [mul_comm, C_mul', f.map_smul, C_eq_smul_one, f.map_smul, H.1, smul_one_smul, smul_zero,
+        smul_zero, zero_add] },
+    case h_add : q₁ q₂ h₁ h₂ {
+      simp only [mul_add, f.map_add, h₁, h₂, smul_add, add_smul], abel },
+    case h_X : q i hq {
+      simp only [this, ← mul_assoc, hq, mul_smul, smul_add, smul_comm (X i), add_assoc] } }
+end
+
+@[simp] lemma derivation_C (f : derivation R (mv_polynomial σ R) A) (a : R) : f (C a) = 0 :=
+f.map_algebra_map a
+
+@[ext] lemma derivation_ext {f g : derivation R (mv_polynomial σ R) A}
+  (h : ∀ i, f (X i) = g (X i)) :
+  f = g :=
+begin
+  apply derivation.to_linear_map_injective, ext s, dsimp,
+  induction s using finsupp.induction with i n s hi hn ihs,
```

Comment:
It worries me that you need this, shouldn't ext be handling this induction for you?

Reply:
No, `f ∘ (λ s, monomial s 1)` is not an object known to `ext`.

Reply:
Ah, I understand now.

Can you prove this about `add_monoid_algebra` first?
```lean
@[ext] lemma add_monoid_algebra.derivation_ext {G} [add_comm_monoid G] [module (add_monoid_algebra R G) A]
  [is_scalar_tower R (add_monoid_algebra R G) A]
  {f g : derivation R (add_monoid_algebra R G) A}
  (h : ∀ i, f (add_monoid_algebra.of' _ _ i) = g (add_monoid_algebra.of' _ _ i)) :
  f = g :=
begin
  apply derivation.to_linear_map_injective,
  ext s,
  exact h s,
end
```
then this becomes
```lean
begin
  ext s, show f (monomial s 1) = g (monomial s 1),
  induction s using finsupp.induction with i n s hi hn ihs,
  { exact f.map_one_eq_zero.trans g.map_one_eq_zero.symm },
  { simp only [monomial_single_add, ihs, h, derivation.leibniz, derivation.leibniz_pow] }
end
```